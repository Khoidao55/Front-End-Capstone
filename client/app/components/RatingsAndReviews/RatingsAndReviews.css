.header {
  margin: 10px 10px 10px 60px;
}

.ratingsAndReviewsContainer {
  display: flex;
  justify-content: center;
}

.breakdownContainer {
  display: flex;
  flex-direction: column;
  justify-content: start;
  width: 400px;
  margin: 0px 10px 10px 10px;
  border: 3px solid #333;
  background-color: #dfdfdf;
  min-height: 500px;
}

.reviewListContainer {
  display: flex;
  flex-direction: column;
  justify-content: start;
  width: 800px;
  margin: 0px 10px 10px 10px;
  border: 3px solid #333;
  background-color: #dfdfdf;
  min-height: 500px;
}

.ratingBreakdown {
  border: 3px solid #333;
  background-color: #dfdfdf;
  min-height: 150px;
  margin: 10px;
}

.productBreakdown {
  border: 3px solid #333;
  background-color: #dfdfdf;
  min-height: 150px;
  margin: 10px;
}

.sortForm {
  border: 3px solid #333;
  background-color: #dfdfdf;
  min-height: 30px;
  margin: 10px;
}

.sortFormLabel {
  margin: 10px 0 0 10px;
}

.reviewList {
  border: 3px solid #333;
  background-color: #dfdfdf;
  min-height: 100px;
  margin: 10px;
}

.buttonsContainer {
  display: flex;
  border: 3px solid #333;
  background-color: #dfdfdf;
  min-height: 100px;
  margin: 10px;
}

.moreReviewsButton {
  border: 3px solid #333;
  background-color: #dfdfdf;
  min-height: 50px;
  margin: 10px;
}

.addReviewButton {
  border: 3px solid #333;
  background-color: #dfdfdf;
  min-height: 50px;
  margin: 10px;
}
