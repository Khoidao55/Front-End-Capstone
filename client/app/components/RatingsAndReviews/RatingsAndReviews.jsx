import React, { useState, useEffect } from 'react';
import axios from 'axios';
import propTypes from 'proptypes';
import './RatingsAndReviews.css';
import RatingSummary from './RatingSummary/RatingSummary.jsx';
import ReviewList from './ReviewList/ReviewList.jsx';
import MoreReviewsButton from './ReviewList/ReviewButtonsContainer/MoreReviewsButton/MoreReviewsButton.jsx';
import AddReviewButton from './ReviewList/ReviewButtonsContainer/AddReviewButton/AddReviewButton.jsx';

const RatingsAndReviews = ({ productID, metaData, productName, setCurrentProductReviews }) => {
  // set up state
  // track current productID as well as the reviews for that product
  const [reviewList, setReviewList] = useState([]);
  const [reviewListStorage, setReviewListStorage] = useState([]);
  const [sortOrder, setSortOrder] = useState('relevant');
  const [visibleReviews, setVisibleReviews] = useState(2);
  const [listIsFiltered, setListIsFiltered] = useState(false);

  // on component mount, use the productID to fetch reviews from the server
  useEffect(() => {
    if (productID) {
      axios.get(`/reviews/sort/${sortOrder}/product/${productID}`) // TODO productID
        .then((response) => {
          setReviewList(response.data.results);
          setReviewListStorage(response.data.results);
          setCurrentProductReviews(response.data);
        })
        .catch((err) => {
          console.log('error fetching data on mount: ', err);
        });
    }
  }, [productID]);

  useEffect(() => {
    if (productID) {
      axios.get(`/reviews/sort/${sortOrder}/product/${productID}`) // TODO productID
        .then((response) => {
          setReviewList(response.data.results);
          setReviewListStorage(response.data.results);
        })
        .catch((err) => {
          console.log('error fetching data on mount: ', err);
        });
    }
  }, [sortOrder]);

  const handleSortClick = (func) => {
    setReviewList(reviewListStorage.filter(func));
    setListIsFiltered(true);
  };

  const handleRemoveFilterClick = (e) => {
    e.preventDefault();
    setReviewList(reviewListStorage);
    setListIsFiltered(false);
  };

  const MoreReviewsButtonRender = () => {
    if (reviewList.length > visibleReviews) {
      return (
        <MoreReviewsButton
          visibleReviews={visibleReviews}
          setVisibleReviews={setVisibleReviews}
        />
      );
    }
    return (
      <></>
    );
  };

  const RemoveAllFiltersRender = () => {
    if (listIsFiltered) {
      return (
        <button type="button" className="remove-review-filters-button" onClick={handleRemoveFilterClick}>Remove all filters</button>
      );
    }
    return (<></>);
  };

  return (
    <>
      <h3 className="reviews-header">RATINGS & REVIEWS</h3>
      <div className="ratings-and-reviews-container">
        <div className="review-breakdown-container">
          <RatingSummary
            className="review-rating-summary"
            metaData={metaData}
            handleSortClick={handleSortClick}
            RemoveAllFiltersRender={RemoveAllFiltersRender}
          />
        </div>

        <div className="review-list-container">
          <form onChange={(event) => { setSortOrder(event.target.value); }} className="sort-form">
            <label htmlFor="sort-form" className="sort-form-label">Sort on </label>
            <select id="sort-form" name="sort-form" className="sort-form-select">
              <option value="relevant">relevant</option>
              <option value="helpful">helpful</option>
              <option value="newest">newest</option>
            </select>
          </form>
          <>
            <ReviewList
              reviewList={reviewList}
              visibleReviews={visibleReviews}
            />
          </>
          <div className="reviewButtonsContainer">
            <MoreReviewsButtonRender />
            <AddReviewButton
              productName={productName}
              productID={productID}
              metaData={metaData}
            />
          </div>
        </div>
      </div>
    </>
  );
};

RatingsAndReviews.propTypes = {
  productID: propTypes.number,
  product: propTypes.object,
  metaData: propTypes.object
};

export default RatingsAndReviews;
