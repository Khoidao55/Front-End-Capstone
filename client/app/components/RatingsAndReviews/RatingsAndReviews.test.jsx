import React from 'react';
import { shallow } from 'enzyme';
import RatingsAndReviews from './RatingsAndReviews';

describe('ratings and reviews', () => {
  test('it renders correctly', () => {
    const wrapper = shallow(<RatingsAndReviews />);

    expect(wrapper).toMatchSnapshot();
  });
});
