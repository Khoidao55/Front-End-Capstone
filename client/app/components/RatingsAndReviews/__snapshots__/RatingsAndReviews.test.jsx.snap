// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ratings and reviews it renders correctly 1`] = `ShallowWrapper {}`;
